pkg/transport/process/util.go

```diff
@@ -0,0 +1,177 @@
+package process
+
+import (
+	"archive/tar"
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"time"
+
+	cdv2 "github.com/gardener/component-spec/bindings-go/apis/v2"
+	"sigs.k8s.io/yaml"
+)
+
+const (
+	componentDescriptorFile = "component-descriptor.yaml"
+	resourceFile            = "resource.yaml"
+	resourceBlobFile        = "resource-blob"
+)
+
+// WriteTARArchive writes the component descriptor, resource and resource blob to a TAR archive
+func WriteTARArchive(ctx context.Context, cd cdv2.ComponentDescriptor, res cdv2.Resource, resourceBlobReader io.Reader, outArchive *tar.Writer) error {
+	defer outArchive.Close()
+
+	marshaledCD, err := yaml.Marshal(cd)
+	if err != nil {
+		return fmt.Errorf("unable to marshal component descriptor: %w", err)
+	}
+
+	err = writeFileToTARArchive(componentDescriptorFile, bytes.NewReader(marshaledCD), outArchive)
+	if err != nil {
+		return fmt.Errorf("unable to write %s: %w", componentDescriptorFile, err)
+	}
+
+	marshaledRes, err := yaml.Marshal(res)
+	if err != nil {
+		return fmt.Errorf("unable to marshal resource: %w", err)
+	}
+
+	err = writeFileToTARArchive(resourceFile, bytes.NewReader(marshaledRes), outArchive)
+	if err != nil {
+		return fmt.Errorf("unable to write %s: %w", resourceFile, err)
+	}
+
+	if resourceBlobReader != nil {
+		err = writeFileToTARArchive(resourceBlobFile, resourceBlobReader, outArchive)
+		if err != nil {
+			return fmt.Errorf("unable to write %s: %w", resourceBlobFile, err)
+		}
+	}
+
+	return nil
+}
+
+func writeFileToTARArchive(filename string, contentReader io.Reader, outArchive *tar.Writer) error {
+	tempfile, err := ioutil.TempFile("", "")
+	if err != nil {
+		return fmt.Errorf("unable to create tempfile: %w", err)
+	}
+	defer tempfile.Close()
```

Comment:
you only need a temporary file for the size?
this seems in that context unnecessary as we already know the size. so you might just add the size as paramter and then we can directly write to the tar

Reply:
for the component descriptor and resource we know the size upfront, but not for the resource blob.

Reply:
so just for simplicity, I would handle everything uniformally, even thought for the component descriptor and resource it's unnecessary as you said

Reply:
ok lets see if we run into perf issues.
An alternative would also be to use a seeked reader. but lets keep it that way for now